rfc/2024-11-13-provider-oci-registry.md

```diff
@@ -0,0 +1,240 @@
+
+- The tag listing endpoint is located at `/v2/<name>/tags/list` and supports additional filtering and pagination parameters. It lists all the tags (a kind of reference) that have manifests.
+- The referrer listing endpoint is located at `/v2/<name>/referrers/<digest>` and returns the list of manifests that refer to a specific blob.
+
+> [!TIP]
+> Try it yourself: authenticate with a token as described above, then use the following command:
+> ```
+> curl -H "Authorization: Bearer djE6b3Blb..." https://ghcr.io/v2/opentofu/opentofu/tags/list
+> ```
+
+### The `_catalog` extension
+
+Although not standardized in the distribution spec, the `/v2/_catalog` endpoint is traditionally used in the [Docker Registry](https://docker-docs.uclv.cu/registry/spec/api/#listing-repositories) as a means to list all images in a registry. However, this endpoint is typically disabled in public registries or only available after authentication.
+
+---
+
+## User Documentation
+
+### For provider authors / mirror operators
+
+There are two possible approaches for distributing providers.
+
+<details>
+<summary>
+
+#### 📂 Standard container tooling approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. This involves the following steps:
+
+1. Take all build artifacts and add them to a container image.
+2. Upload the container image to an OCI registry.
+
+For example, assuming you have your binary built and wanted to package it up into a container, you could use the following `Dockerfile`:
+
+```Dockerfile
+FROM scratch
+
+ADD terraform-provider-aws
+ADD README.md
+ADD LICENSE
+ADD CHANGELOG.md
+```
+
+You could now run `docker build -t ghcr.io/opentofu/terraform-provider-aws:1.2.3-amd64 .` followed by `docker push ghcr.io/opentofu/terraform-provider-aws:1.2.3`.
+
+For more advanced use cases, you could enable container image building in your [goreleaser config](https://goreleaser.com/customization/docker/) to enable building multi-arch images.
+</details>
+
+<details>
+<summary>
+
+#### 📂 ORAS approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. You can do this by using the [ORAS command line tool](https://oras.land/docs/installation) using the following commands:
+
+```
+oras login -u username -p password ghcr.io
+oras push ghcr.io/opentofu/terraform-provider-aws:1.2.3 ./dist:application/octet-stream+tar
+```
+
+🛑🛑🛑 TODO: how do you create multi-arch with ORAS? 🛑🛑🛑
+
+</details>
+
+### Mirror configuration
+
+If you would like to use an OCI registry as your OpenTofu mirror, you will be able to do so by specifying a `provider_installation` block with an `oci_mirror` section as follows:
+
+```hcl
+provider_installation {
+  oci_mirror {
+    url = "ghcr.io/{{ .Namespace }}/terraform-provider-{{ .Name }}"
```

Comment:
This would be the first example of using the Go template language for something in core OpenTofu, so it feels a little inconsistent.

I think we should try to use HCL-style templates here instead, for consistency. We don't generally assume that OpenTofu users are familiar with any part of the Go ecosystem. That is admittedly a little challenging in this context because the CLI configuration still uses legacy HCLv1 rather than HCLv2 that we use everywhere else, but one way to do it would be to pass the resulting string to [`hclsyntax.ParseTemplate`](https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclsyntax#ParseTemplate) to get an effect similar to how interpolation worked in older versions before HCLv2[^1].

[^1]: HCLv1's parser has a special case where it recognizes substrings starting with `${` and ending with `}` as having looser escaping rules, because it was made to work with [HIL](https://github.com/hashicorp/hil). HCLv2's template syntax is similar enough to HIL that the same special case should work for HCLv2 templates too.

Since we're talking about provider _mirrors_ we should presumably also include `hostname` as a valid variable, because otherwise this installation rule would generate the same template regardless of which hostname is selected, causing the different namespaces to become muddled.

If someone wants to write an `oci_mirror` rule that is for only one hostname then they could presumably use the `include` and `exclude` arguments we support for the other installation methods, and then they would not need to include `hostname` in their template.

```hcl
provider_installation {
  oci_mirror {
    url     = "ghcr.io/${namespace}/terraform-provider-${name}"
    include = ["registry.opentofu.org/*/*"]
  }
}
```

We could perhaps guide users to a correct configuration here by checking that the template includes a reference to each component that is wildcarded in `include`, and returning an error/warning if not. For example, since `include` has the namespace wildcarded above, it would be required to refer to `namespace` at least once in the template, but this template doesn't need to include `hostname` because that isn't wildcarded. The default `include` is `["*/*/*"]` so all three would be required if `include` isn't set at all.

Reply:
The string that results from this template also doesn't use standard URL syntax, so perhaps we should choose an argument name other than `url` to refer to it. Is there a conventional name for this shorthand address form in the Docker community?